File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Delete_the_Middle_Node_of_a_Linked_List/yujin.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        n = 0
        cur = head
        while cur:
            n += 1
            cur = cur.next
        middle = n // 2
        if n == 1:
            head = None
        elif n == 2:
            head.next = None
        else:
            index = 0
            cur = head
            while index < middle:
                if index == middle-1:
                    cur.next = cur.next.next
                cur = cur.next
                index += 1

        return head

File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Delete_the_Middle_Node_of_a_Linked_List/yujin.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        int n = 0;
        ListNode* cur = head;
        while (cur != nullptr) {
            cur = cur->next;
            n ++;
        }
        if (n == 1) {
            head = nullptr;
        }
        else if (n == 2)
        {
            head->next = nullptr;
        }
        else {
            int index = 0;
            int middle = n / 2;
            ListNode* cur = head;
            while (index < middle) {
                if (index == middle - 1) {
                    cur->next = cur->next->next;
                }
                cur = cur->next;
                index ++;
            }
        }
        return head;
    }
};
